### PR TITLE
Port perf changes and issues

### LINKED ISSUE #40449: Possible unintended NullReferenceException thrown from JsonSerializer.Deserialize
Tested with master branch (as of 2fac89f63c644ab81d936d62c91e6003302e718d)
The following code:

```c#
using System.Text.Json;
using System.Collections.Generic;

interface X {}

class A : List<X> {}

class Program
{
  static void Main(string[] args) => JsonSerializer.Deserialize<A>("false");
}
```

throws `NullReferenceException`:

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Text.Json.JsonPropertyInfo.Read(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.HandleValue(JsonTokenType tokenType, JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
```

if we remove `: List<X>`, then it throws `JsonException`:

```
Unhandled exception. System.Text.Json.JsonException: The JSON value could not be converted to myApp.A. Path: $ | LineNumber: 0 | BytePositionInLine: 5.
   at System.Text.Json.ThrowHelper.ThrowJsonException(String message, Utf8JsonReader& reader, String path, Exception innerException)
   at System.Text.Json.ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type propertyType, Utf8JsonReader& reader, String path, Exception innerException)
   at System.Text.Json.JsonPropertyInfoNotNullable`4.OnRead(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonPropertyInfo.Read(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.HandleValue(JsonTokenType tokenType, JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
```

if we keep `: List<X>` and replace `"false"` with `"{}"`, then we get `NotSupportedException`:

```
Unhandled exception. System.NotSupportedException: Deserialization of interface types is not supported. Type 'X'
   at System.Text.Json.ThrowHelper.ThrowInvalidOperationException_DeserializePolymorphicInterface(Type invalidType)
   at System.Text.Json.JsonSerializer.HandleStartObject(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
```

I think `NotSupportedException` is desired in place of NRE from the original repro until there is an interface support. After the support is added, `JsonException` would be desired here (similar to the one we get without `: List<X>`).

### LINKED ISSUE #40704: JsonSerializer always encodes dictionary keys and property names
I have some code as below:
``` csharp
var dict = new Dictionary<string, Dictionary<string, string>>();
dict.Add("用户组01", new Dictionary<string, string>
{
    { "用户ABC","123"},
    { "用户DEF","456"}
});
var json = JsonSerializer.Serialize(dict);
```
and got the json as below:
```
{"\u7528\u6237\u7EC401":{"\u7528\u6237ABC":"123","\u7528\u6237DEF":"456"}}
```
when I actually want:
```
{"用户组01":{"用户ABC":"123","用户DEF":"456"}}
```

### LINKED ISSUE #41425: JsonConverterAttribute can't be used to deserialize IEnumerable properties
When trying to deserialize an object where one of the properties is an IEnumerable/collection and has a custom converter set by `JsonConverterAttribute`, a `JsonException` is thrown. Repro:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace JsonConverterRepro
{
    class Program
    {
        static void Main(string[] args)
        {
            var options = new JsonSerializerOptions();
            // Program works as expected when this line is uncommented:
            // options.Converters.Add(new DummyConverter());

            var json = @"{ ""Items"": [""hello"", 1, true] }";
            var model = JsonSerializer.Deserialize<MyModel>(json, options);

            Console.WriteLine(string.Join(", ", model.Items));
        }

        class MyModel
        {
            [JsonConverter(typeof(ToStringConverter))]
            public List<string> Items { get; set; }
        }

        // This converter coerces list items into strings when deserializing
        class ToStringConverter : JsonConverter<List<string>>
        {
            public override List<string> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
            {
                using var doc = JsonDocument.ParseValue(ref reader);
                return doc.RootElement.EnumerateArray().Select(e => e.ToString()).ToList();
            }

            public override void Write(Utf8JsonWriter writer, List<string> value, JsonSerializerOptions options)
            {
                JsonSerializer.Serialize(writer, value, options);
            }
        }

        class DummyConverter : JsonConverter<List<string>>
        {
            public override List<string> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
                throw new NotImplementedException();

            public override void Write(Utf8JsonWriter writer, List<string> value, JsonSerializerOptions options) =>
                throw new NotImplementedException();
        }
    }
}
```

This program should output `hello, 1, True`. However, when this is actually run, this exception is thrown:
```
Unhandled exception. System.Text.Json.JsonException: The JSON value could not be converted to System.Collections.Generic.List`1[System.String]. Path: $.Items | LineNumber: 0 | BytePositionInLine: 12.
 ---> System.InvalidOperationException: Cannot get the value of a token type 'StartArray' as a string.
   at System.Text.Json.Utf8JsonReader.GetString()
   at System.Text.Json.Serialization.Converters.JsonConverterString.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
   at System.Text.Json.JsonPropertyInfoNotNullable`4.OnReadEnumerable(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonPropertyInfo.ReadEnumerable(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonPropertyInfo.Read(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.HandleValue(JsonTokenType tokenType, JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& readStack, Utf8JsonReader& reader, Exception ex)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at JsonConverterRepro.Program.Main(String[] args) in H:\projects\JsonConverterRepro\Program.cs:line 17
```

The correct behaviour is observed when when a custom converter for that property's type is added to the `JsonSerializerOptions`, even if that custom converter doesn't get used. This program also works correctly when the `ToStringConverter` is added to the `JsonSerializerOptions` instead of being set via an attribute.

Although this repro uses `List<string>`, the same problem occurs with other collection types and item types e.g. `IEnumerable<string>`, `HashSet<int>`.

Tested on .NET Core 3.0.100

### PR DESCRIPTION
**Pull request by steveharter**
The "No Merge" label is there to ensure when this is merged,the individual commits are NOT squashed (one commit here per original commit in master).

Peformance and three bug fixes:
Fixes #40449
Fixes #40704 
Fixes #41425 

On performance:
•	Serialization is ~1.2x to ~1.5x faster than 3.0
•	Deserialization is ~1.2x faster than 3.0

Ports the following:

Issue | Type | 5.0 PR | 5.0 Commit | Description | Customer Impact | Risk
-- | -- | -- | -- | -- | -- | --
40449 | Bug | 40501 | 5dbbdfa | Change a NullReferenceException to   JsonException when invalid JSON is encountered in certain cases. | Reported by the community. Invalid   JSON can result in a NullReferenceException which is not expected and likely   not in any Try\Catch. | Low. The fix has been in master   for several weeks. Not likely to be “breaking” since changing from   NullReferenceException to another (valid) exception is not considered   breaking.
40704 | Bug | 40787 | 7829d4a | Non-ASCII dictionary key names and   object property names are always escaped. | Reported by the community.   Non-ASCII locales have additional JSON payload size and in general this is   not expected. | Low. The fix has been in master   for several weeks. Appropriate tests added.
  | Perf | 40889 | 27ddbe9 | Minor deserialization perf   improvements for collection by removing unnecessary allocations. | (perf) | Low. The fix has been in master   for several weeks.
  | Perf | 40998 | d3c6628 | Major perf deserialzation perf   gains primarily due to changes to property name lookup. | (perf) | Low. The fix has been in master   for several weeks. Appropriate tests added.
  | Perf | 41098 | 017a038 | Major serialization perf   improvements based on [AggressiveInlining] and enumerator changes. | (perf) | Low. The fix has been in master   for several weeks.
  | Perf | 41238 | 527595f | Medium perf gains on deserialize   and serialize using a variety of changes. | (perf) | Low. The fix has been in master   for two weeks. Appropriate tests added.
  | Perf | 41363 | 4cfc7ec | Minor perf gains with the   JsonCamelCaseNamingPolicy. This decreases warmup time but can help at runtime   when using the naming policy on Dictionaries. | (perf) | Low. The fix has been in master   for two weeks.
  | Perf | 41414 | 8655ef9 | Minor perf gains on   deserialization; includes some clean-up and addresses late comments from   previous PR. | (perf) | Low. Appropriate tests added.
41425 | Bug | 41653 | 4580a4d | A [JsonConverter] attribute when   applied to a collection property to specify a custom converter does not work   (does nothing). | Reported by the community. Having   a broken feature is confusing\not expected when encountered. Note there is a   work-around to add the converter at run-time (instead of an attribute). | Low. The fix is contained in a   single method and new tests added.

### CODE REVIEW

**Comment by ahsonkhan** [comment]
How much effort would it be to port https://github.com/dotnet/corefx/pull/40867 as well? I understand that the code changed since then, but it would be good to keep that history of changes in sync.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs [hunk 1]
```diff
@@ -159,7 +159,8 @@ public static partial class JsonSerializer
                     }
                 }
 
-                writer.WritePropertyName(key);
+                JsonEncodedText escapedKey = JsonEncodedText.Encode(key, options.Encoder);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: This is an incorrect change. It is undoing the optimizations from https://github.com/dotnet/corefx/pull/40991#discussion_r324953747 / https://github.com/dotnet/corefx/pull/41210

Please verify that nothing else from that PR is being changed, unintentionally.

**Comment by steveharter** [comment]
> How much effort would it be to port #40867 as well? 

The code that was changed with #40867 was replaced with https://github.com/dotnet/corefx/pull/41098 (included here)

**Comment by steveharter** [comment]
> Also, were the commits here applied in the same order as they were checked-in to master?

Yes

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: Thanks! Updated that and re-verified other commits.

Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: > Also, how did this diff fall through?

It didn't. That fix was done in an earlier change and was out-of-order and manually applied.

**Comment by steveharter** [comment]
I did a manual diff between 5.0 and this PR and found this perf PR (https://github.com/dotnet/corefx/pull/40114) that I just adorned with "performance" label.

It has minor perf benefits, but is the only current 5.0 perf issue AFAIK not ported here. We should consider adding here (or in a subsequent PR).

**Comment by ericstj** [comment]
We can do in subsequent PR.  I've reviewed all the commits and am OK with bringing them over to 3.1.  We just need to document https://github.com/dotnet/corefx/pull/40787.

**Comment by danmosemsft** [comment]
As far as I/tactics are concerned, this can be merged.

BTW if someone accidentally squashes, I suggest reversing that commit and then doing it again so we have that nice history. I only mention this because I can totally imagine myself doing this 😄

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterKeyValuePair.cs [hunk 2]
```diff
@@ -11,8 +11,11 @@ internal sealed class JsonKeyValuePairConverter<TKey, TValue> : JsonConverter<Ke
         private const string KeyName = "Key";
         private const string ValueName = "Value";
 
-        private static readonly JsonEncodedText _keyName = JsonEncodedText.Encode(KeyName);
-        private static readonly JsonEncodedText _valueName = JsonEncodedText.Encode(ValueName);
+        // "encoder: null" is used since the literal values of "Key" and "Value" should not normally be escaped
+        // unless a custom encoder is used that escapes these ASCII characters (rare).
+        // Also by not specifying an encoder allows the values to be cached statically here.
+        private static readonly JsonEncodedText _keyName = JsonEncodedText.Encode(KeyName, encoder: null);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterKeyValuePair.cs]: Out of scope of this PR, but we should pass in the user-defined encoder for this (and not null). Though unlikely, if a user creates an encoder that escapes everything, then the payload would be incorrect.

**Pull request merged**